ambry-cloud/src/main/java/com.github.ambry.cloud/CloudBlobStore.java

```diff
@@ -114,7 +114,85 @@ public void start() throws StoreException {
+      Map<String, CloudBlobMetadata> cloudBlobMetadataListMap = cloudDestination.getBlobMetadata(blobIdList);
+      if (cloudBlobMetadataListMap.size() < blobIdList.size()) {
+        Set<BlobId> missingBlobs = blobIdList.stream()
+            .filter(blobId -> !cloudBlobMetadataListMap.containsKey(blobId))
+            .collect(Collectors.toSet());
+        throw new StoreException("Some of the keys were missing in the cloud metadata store: " + missingBlobs,
+            StoreErrorCodes.ID_Not_Found);
+      }
+      validateCloudMetadata(cloudBlobMetadataListMap, storeGetOptions);
+      for (BlobId blobId : blobIdList) {
+        CloudBlobMetadata blobMetadata = cloudBlobMetadataListMap.get(blobId.getID());
+        MessageInfo messageInfo = new MessageInfo(blobId, blobMetadata.getSize(), blobMetadata.getExpirationTime(),
+            (short) blobMetadata.getAccountId(), (short) blobMetadata.getContainerId(), getOperationTime(blobMetadata));
+        messageInfos.add(messageInfo);
+        blobReadInfos.add(new CloudMessageReadSet.BlobReadInfo(blobMetadata, blobId));
+      }
+    } catch (CloudStorageException e) {
+      throw new StoreException(e, StoreErrorCodes.IOError);
+    }
+    CloudMessageReadSet messageReadSet = new CloudMessageReadSet(blobReadInfos, cloudDestination);
+    return new StoreInfo(messageReadSet, messageInfos);
+  }
+
+  /**
+   * validate the {@code CloudBlobMetadata} map to make sure it has metadata for all keys, and they meet the {@code storeGetOptions} requirements.
+   * @param cloudBlobMetadataListMap
+   * @throws StoreException if the {@code CloudBlobMetadata} isnt valid
+   */
+  private void validateCloudMetadata(Map<String, CloudBlobMetadata> cloudBlobMetadataListMap,
+      EnumSet<StoreGetOptions> storeGetOptions) throws StoreException {
+    for (String key : cloudBlobMetadataListMap.keySet()) {
+      if (isBlobDeleted(cloudBlobMetadataListMap.get(key)) && !storeGetOptions.contains(
+          StoreGetOptions.Store_Include_Deleted)) {
+        throw new StoreException("Id " + key + " has been deleted on the cloud", StoreErrorCodes.ID_Deleted);
+      }
+      if (isBlobExpired(cloudBlobMetadataListMap.get(key)) && !storeGetOptions.contains(
+          StoreGetOptions.Store_Include_Expired)) {
+        throw new StoreException("Id " + key + " has expired on the cloud", StoreErrorCodes.TTL_Expired);
+      }
+    }
+  }
+
+  /**
+   * Gets the operation time for a blob from blob metadata based on the blob's current state and timestamp recorded for that state.
+   * @param metadata blob metadata from which to derive operation time.
+   * @return operation time.
+   */
+  private long getOperationTime(CloudBlobMetadata metadata) {
+    if (isBlobDeleted(metadata)) {
+      return metadata.getDeletionTime();
+    } else if (isBlobExpired(metadata)) {
+      return metadata.getExpirationTime();
+    }
+    return metadata.getCreationTime();
```

Comment:
Unfortunately, many blobs uploaded to Azure have creation time = -1.  (No way you could have known that.) . Please check for this and use upload time as backup.
I'm not sure how the operation time is being used downstream, but what we have in CloudBlobMetadata may not be that meaningful.  For example, deletion time isn't the same as operation time (time deletion was issued), and expiration time is set separately in the MessageInfo.  Need to think more about this.

Reply:
Is there any reason why we have blobs uploaded to Azure with a different creation time than the creation time reported from the originating blobstore partition?
Also is there a reason we don't have "operationTime" in CloudBlobMetadata? If we have an additional field called "operationTime" in the CloudBlobMetadata (similar to one we have in the index metadata on stores), then we can use the following rules to populate operationTime on CloudBlobMetadata:

1. If the blob is replicated or backed-up to a replica (irrespective of the destination being cloud or not), the operation time should just be copied from what is present in the original data.
2. Once we are serving both get and put from azure, we can then use the actual time of operation as the "operationTime". This when replicated to other Azure colos or to linkedin replicas, will follow the rule 1 above.

Unless ofcourse I am missing something here.